File: src/ctx/gitinfo.py
from __future__ import annotations

import subprocess
from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class GitInfo:
    sha: str
    dirty: bool


def git_info(root: Path) -> GitInfo:
    try:
        sha = subprocess.run(
            ["git", "rev-parse", "--short", "HEAD"], cwd=root,
            text=True, capture_output=True, check=True, timeout=2,
        ).stdout.strip()
        dirty = bool(subprocess.run(
            ["git", "status", "--porcelain"], cwd=root,
            text=True, capture_output=True, check=True, timeout=3,
        ).stdout.strip())
        return GitInfo(sha=sha, dirty=dirty)
    except (FileNotFoundError, subprocess.SubprocessError):
        return GitInfo(sha="nogit", dirty=False)

File: src/gitinfo.rs
use std::path::Path;
use std::process::Command;

use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct GitInfo {
    pub sha: String,
    pub dirty: bool,
}

pub fn git_info(root: &Path) -> GitInfo {
    let sha = git_output(root, &["rev-parse", "--short", "HEAD"])
        .filter(|value| !value.is_empty())
        .unwrap_or_else(|| "nogit".to_owned());
    let dirty = git_output(root, &["status", "--porcelain"])
        .map(|value| !value.is_empty())
        .unwrap_or(false);
    GitInfo { sha, dirty }
}

fn git_output(root: &Path, arguments: &[&str]) -> Option<String> {
    let output = Command::new("git")
        .args(arguments)
        .current_dir(root)
        .output()
        .ok()?;
    if !output.status.success() {
        return None;
    }
    Some(String::from_utf8_lossy(&output.stdout).trim().to_owned())
}
